src/main/java/org/cryptomator/cloudaccess/requestdecorator/QuotaRequestCachingDecorator.java

```diff
@@ -0,0 +1,47 @@
+package org.cryptomator.cloudaccess.requestdecorator;
+
+import com.github.benmanes.caffeine.cache.AsyncCache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import org.cryptomator.cloudaccess.api.CloudPath;
+import org.cryptomator.cloudaccess.api.CloudProvider;
+import org.cryptomator.cloudaccess.api.Quota;
+import org.cryptomator.cloudaccess.api.exceptions.NotFoundException;
+import org.cryptomator.cloudaccess.api.exceptions.QuotaNotAvailableException;
+
+import java.time.Duration;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * Decorates an existing CloudProvider by caching quota-requests for a duration of default 10 seconds (can be set using <code>org.cryptomator.cloudaccess.metadatacachingprovider.timeoutSeconds</code>).
+ */
+class QuotaRequestCachingDecorator extends CloudProviderDecorator {
+
+	private final static int DEFAULT_CACHE_TIMEOUT_SECONDS = 10;
+
+	// visible for testing
+	final AsyncCache<CloudPath, Quota> quotaCache;
+
+	private final CloudProvider delegate;
+
+	public QuotaRequestCachingDecorator(CloudProvider delegate) {
+		this(delegate, Caffeine
+				.newBuilder()
+				.expireAfterWrite(Duration.ofSeconds(Integer.getInteger("org.cryptomator.cloudaccess.metadatacachingprovider.timeoutSeconds", DEFAULT_CACHE_TIMEOUT_SECONDS)))
```

Comment:
should we adjust the property name to reflect that it only affects quota requests?

Reply:
We could but it is used in other decorators too so I wouldn't recommend it because it fits with this context quite okay IMO :sweat_smile:  (not good that is true).

Reply:
all right